File: DesignTool/docs/inOutChecker.js
function InOutChecker(region) {
	//
	// Check is a point is inside region
	// Points is a list of (x,y, flag) 
	// Region is a list of points
	// flag will be set if inside, cleared if outside
	
	var self = this;
	
	function buildEdgeList(lines) {
		function addEdge (el, x1, y1, x2, y2) {
			if (y1 == y2) return;
			if (y1 > y2) {
				var t = x1;				
				x1 = x2;
				x2 = t;
				t = y1; 
				y1 = y2;
				y2 = t;				
			}			

			var m, b, x, y;
			m = (x2-x1) / (y2-y1);
			b = -m * y1 + x1 
			for (y = y1; y < y2; ++y) {
				x = y * m + b;
				if (!el[y])
					el[y] = [];
				el[y].push (x);
			}
		}
		var edgeList = [];
		var i;
		var x0, y0;
		var lastX, lastY;
		var x, y, code;
		var row;
		var ymin = 1E10, ymax = -99999;
		for (i in lines) {
			row = lines[i];
			//print ("line ", i, row);
			x = row[0];
			y = row[1];
			code = row[2];
			y = Math.floor(y);
			ymin = Math.min (ymin, y);
			ymax = Math.max (ymax, y);
			if (code == 0) {
				// start vertex
				lastX = x0 = x;
				lastY = y0 = y;
				continue;
			}
			if (code == 1) {
				// next vertex
				// edge = x,y -> lastX, lastY
				addEdge (edgeList, x, y, lastX, lastY);
				lastX = x;
				lastY = y;
				continue;
			}
			if (code == 2) {
				// end vertex, should be same as start vertex
				// edge = x,y -> x0, y0
				addEdge (edgeList, lastX, lastY, x0, y0);
			}
		}
		for (y = ymin; y < ymax; ++y) {
			if (edgeList[y])
				edgeList[y].sort(function(a, b) {return a-b;});
		}
		return edgeList;
	}		
	
	self.checkPoint = function (x, y) {
		var row = self.edgeList[Math.floor(y)];
		if (!row) return false;
		
		var i, len;
		len = row.length;
		len = len - (len % 2);
		i = 0;
		while (i < len) {
			if ((row[i] < x) && (x < row[i+1])) return true;
			i += 2;
		}
		return false;		
	};
	
	self.edgeList = buildEdgeList (region);
	
	return self;
}

File: DesignTool/smdtLibs/inOutChecker.py
"""
In/Out checker, python version

Example:

chk = InOutChecker (mask)
pnt = x, y
if chk.checkPoint (x, y):
    print ("inside")

Date: 2018-08-03
Author: Shui Hung Kwok

"""
import math


class InOutChecker:
    def __init__(self, mask):
        """
        mask defines a set of polygons specified as a list of points. 
        A point is a tuple (x, y, flag), 
        where flag is 0 (move), 1 (goto), 2 (back to first point).
        """
        self.mask = mask
        self.ymin, self.ymax, self.edges = self._buildEdges(mask)

    def _buildEdges(self, mask):
        """
        Builds a list of segments from ymin to ymax.
        For each y, a list of segments define what is inside.
        If a point is within a segment, then it is inside.        
        """
        x0, y0 = 0, 0
        x1, y1, x2, y2 = 0, 0, 0, 0
        ys = [int(t[1]) for t in mask]
        ymin, ymax = min(ys), max(ys)
        edges = {}
        for x, y, flag in mask:
            if flag == 0:
                x0, y0 = x, y
                x1, y1 = x, y
                x2, y2 = x, y
                continue
            elif flag == 1:
                x1, y1 = x2, y2
                x2, y2 = x, y
            elif flag == 2:
                x1, y1 = x2, y2
                x2, y2 = x0, y0

            if y1 > y2:
                xx1, yy1, xx2, yy2 = x2, y2, x1, y1
            else:
                xx1, yy1, xx2, yy2 = x1, y1, x2, y2

            y1i, y2i = int(yy1), int(yy2)
            if y1i == y2i:
                continue
            # print (y1i, y2i)
            m = (xx2 - xx1) / (yy2 - yy1)
            b = -m * yy1 + xx1
            for yidx in range(y1i, y2i):
                xpos = m * yidx + b
                try:
                    edges[yidx].append(xpos)
                except:
                    edges[yidx] = [xpos]

        for y in range(ymin, ymax):
            try:
                edges[y].sort()
            except Exception as e:
                print("Exception in sort ", y, e)
                pass
        return ymin, ymax, edges

    def checkPoint(self, x, y):
        """
        Checks if the given point (x,y) is inside the mask
        First, finds the segments at the y position, where int(y) is used.
        Then, the point is inside the mask if x in inside one of the segments. 
        """
        try:
            yi = math.floor(y)
            if yi <= self.ymin or self.ymax <= yi:
                return False
            row = self.edges[yi]

            for idx in range(0, len(row), 2):
                x0, x1 = row[idx : idx + 2]
                if x0 < x < x1:
                    return True
        except Exception as e:
            print("Check point exception", e)
            pass
        return False
